### PR TITLE
Jupiter desktop integration for Solana swaps

### PR DESCRIPTION
**Pull request by onyb**
Resolves https://github.com/brave/brave-browser/issues/19474
Resolves https://github.com/brave/brave-browser/issues/24559

|Security|https://github.com/brave/security/issues/960|
-|-

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [x] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [x] New files have MPL-2.0 license header
- [x] Adequate test coverage exists to prevent regressions
- [x] Major classes, functions and non-trivial code blocks are well-commented
- [x] Changes in component dependencies are properly reflected in `gn`
- [x] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [x] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Demo

https://user-images.githubusercontent.com/3684187/180253987-5ee5b8f1-6e22-44c8-9294-db5645016913.mov

### CODE REVIEW

**Review by Douglashdaniel — APPROVED** [review]
++ Front-end

**Comment by onyb** [comment]
Browser tests crashing - could be related to https://github.com/brave/brave-browser/issues/24129.

File: components/brave_wallet/browser/solana_tx_manager.cc
Comment (yrliou) [inline components/brave_wallet/browser/solana_tx_manager.cc:369]: This should probably be named as `MakeTxDataFromBase64EncodedMessage`.

File: components/brave_wallet/browser/solana_tx_manager.cc
Comment (yrliou) [inline components/brave_wallet/browser/solana_tx_manager.cc:385]: This shouldn't be used if what you have is just message bytes. This one is for signed transactions. You should use SolanaMessage::Deserialize and create your transaction from it.

File: components/brave_wallet/browser/solana_tx_manager.cc
Comment (yrliou) [inline components/brave_wallet/browser/solana_tx_manager.cc:387]: Why do we clear out their blockhash?

File: components/brave_wallet/browser/solana_tx_manager.cc
Comment (yrliou) [inline components/brave_wallet/browser/solana_tx_manager.cc:372]: Is this being used by Jupiter?

**Review by yrliou — CHANGES_REQUESTED** [review]

File: components/brave_wallet/browser/solana_tx_manager.cc
Comment (onyb) [reply components/brave_wallet/browser/solana_tx_manager.cc:385]: The base64-encoded string provided by Jupiter is actually signed transaction bytes (compact-array of signatures followed by message). In this particular case, the signature is empty (all zeros). I think the reason is that Jupiter expects clients to do the following:

```js
import { Transaction } from '@solana/web3.js'
Transaction.from(Buffer.from(serializedTransaction, 'base64'))
```

With that in mind, maybe we should rename our function to `MakeTxDataFromBase64EncodedTranasction`.

File: components/brave_wallet/browser/solana_tx_manager.cc
Comment (onyb) [reply components/brave_wallet/browser/solana_tx_manager.cc:372]: Yes, Jupiter requires us to set `skipPreflight` to `true`.

File: components/brave_wallet/browser/solana_tx_manager.cc
Comment (onyb) [reply components/brave_wallet/browser/solana_tx_manager.cc:369]: Renamed to `MakeTxDataFromBase64EncodedTransaction`.

File: components/brave_wallet/browser/solana_tx_manager.cc
Comment (onyb) [reply components/brave_wallet/browser/solana_tx_manager.cc:387]: I was falsely making the assumption that we should fetch the latest blockhash and use that value whenever the transaction will be approved. We don't rely on such assumptions, so I have removed it.

File: components/brave_wallet/browser/solana_tx_manager.cc
Comment (onyb) [reply components/brave_wallet/browser/solana_tx_manager.cc]: It wasn't necessary - removed it here: fb1676e62c46da736742644bc55d91449423f6db.

File: components/brave_wallet/common/brave_wallet.mojom
Comment (onyb) [reply components/brave_wallet/common/brave_wallet.mojom]: Done here: e97b884fb35322de7f4bc834feb56bbdd295b0dd

File: components/brave_wallet_ui/common/hooks/swap.ts
Comment (onyb) [reply components/brave_wallet_ui/common/hooks/swap.ts:510]: This is was the setup and cleanup transactions, not the actual swap. I used this for lack of a better alternative. I can use `Other` if you think it's more appropriate in this case.

File: components/brave_wallet/browser/fil_tx_manager_unittest.cc
Comment (onyb) [reply components/brave_wallet/browser/fil_tx_manager_unittest.cc]: Fixed here: e67a05fe66c7ba7466d6514f6378e41de48c27b5

File: components/brave_wallet/browser/eth_tx_manager_unittest.cc
Comment (onyb) [reply components/brave_wallet/browser/eth_tx_manager_unittest.cc]: Fixed here: e67a05fe66c7ba7466d6514f6378e41de48c27b5

File: components/brave_wallet/browser/eth_tx_manager_unittest.cc
Comment (onyb) [reply components/brave_wallet/browser/eth_tx_manager_unittest.cc]: Fixed here: e67a05fe66c7ba7466d6514f6378e41de48c27b5

File: components/brave_wallet/browser/solana_transaction.cc
Comment (onyb) [reply components/brave_wallet/browser/solana_transaction.cc:365]: Fixed here: e6e6ecc1b0c96844acaf4e91d808be9582154483

File: components/brave_wallet_ui/common/hooks/swap.ts
Comment (onyb) [reply components/brave_wallet_ui/common/hooks/swap.ts]: Changed to `TransactionType.Other` (58caaf18e268c7af18c8a71bcc41f3b142fbddd0 and 43ef72be80b9afb1c2f5461599c98158b9b85980) assuming you're ok with it. 

Updated screenshot below (cc: @Douglashdaniel for review):
Screenshot 2022-07-27 at 23 48 11

File: components/brave_wallet_ui/utils/random-utils.ts
Comment (onyb) [reply components/brave_wallet_ui/utils/random-utils.ts]: Fixed here: 40f6fa5a255a15e07206b3fee1fc226ffa4af0db

File: components/brave_wallet_ui/common/hooks/swap.ts
Comment (diracdeltas) [inline components/brave_wallet_ui/common/hooks/swap.ts:762]: do you have a screenshot of this? cc @kdenhartog

Comment (diracdeltas) [reply components/brave_wallet_ui/common/hooks/swap.ts:762]: oh is it the screenshot in https://github.com/brave/security/issues/960 ? if so NVM

Comment (kdenhartog) [reply components/brave_wallet_ui/common/hooks/swap.ts:762]: Will give this a look to understand the tradeoff here a bit more and let you know what I think. Thanks for tagging me

File: components/brave_wallet/browser/solana_tx_manager.cc
Comment (onyb) [reply components/brave_wallet/browser/solana_tx_manager.cc:382]: `transaction_bytes` is `nullptr` if the encoded transaction string is an invalid base64 message.  I believe this is covered in the if-statement above:

```cpp
if (!transaction_bytes || transaction_bytes->empty() ||
      transaction_bytes->size() > kSolanaMaxTxSize) { ... }
```

This case is also covered in unit-tests. Does that address your concern?

File: components/brave_wallet/browser/swap_request_helper.cc
Comment (onyb) [reply components/brave_wallet/browser/swap_request_helper.cc:29]: In practice, this condition will never be met as fees are collected in associated token accounts. The only possibility I can think of is when `output_mint` is invalid, in which case we default to `kSolanaFeeRecipient`.

File: components/brave_wallet/browser/solana_tx_manager.cc
Comment (onyb) [reply components/brave_wallet/browser/solana_tx_manager.cc]: Added here: de66ca07d77ca57ac8f2862a1801fdfcb7e583ee

File: components/brave_wallet/browser/fil_tx_manager_unittest.cc
Comment (onyb) [reply components/brave_wallet/browser/fil_tx_manager_unittest.cc]: Fixed it here (5695837baa17167697512c5c1ac680a05bd473ae) and for all other cases too. Going to resolve your other comments except this one.

File: components/brave_wallet_ui/common/hooks/swap.ts
Comment (onyb) [reply components/brave_wallet_ui/common/hooks/swap.ts:762]: Screenshot 2022-08-04 at 20 42 12

File: components/brave_wallet/browser/solana_tx_manager.cc
Comment (yrliou) [reply components/brave_wallet/browser/solana_tx_manager.cc:382]: No, even it's a valid base64 string doesn't mean SolanaTransaction::FromSignedTransactionBytes would success, we can't assume it's going to be a valid encoded transaction that can be deserialized successfully. It's not the same case you tested, need new cases where FromSignedTransactionBytes would return nullptr.

File: components/brave_wallet/browser/swap_request_helper.cc
Comment (yrliou) [reply components/brave_wallet/browser/swap_request_helper.cc:29]: What's the reason to default to that rather than consider it an error?

File: components/brave_wallet/browser/solana_transaction_unittest.cc
Comment (yrliou) [inline components/brave_wallet/browser/solana_transaction_unittest.cc:318]: `const std::vector<uint8_t> empty_signatures(128, 0);`

File: components/brave_wallet/browser/solana_transaction_unittest.cc
Comment (yrliou) [inline components/brave_wallet/browser/solana_transaction_unittest.cc:357]: Aside from this test case, I'm curious of how real tx returned by Jupiter usually look like in terms of signatures, could you share so we can discuss if there are other cases we want to make sure we have the coverage? Thanks.

Comment (onyb) [reply components/brave_wallet/browser/solana_transaction_unittest.cc]: As discussed in DM, I changed the test to use a transaction with single signature in order to better reflect the reality with Jupiter.

a6612754dd3093e1d1d5b608f74c012cb74ba56a

File: components/brave_wallet/browser/solana_tx_manager.cc
Comment (onyb) [reply components/brave_wallet/browser/solana_tx_manager.cc:382]: Sorry I was looking at the wrong line. Fixed it here: 99580b93c3ba6370d718b7c904f1b3c64191bba8.

File: components/brave_wallet/browser/swap_request_helper.cc
Comment (onyb) [reply components/brave_wallet/browser/swap_request_helper.cc]: No reason really - the transaction would fail anyway on-chain, so better to fail fast. Fixed here:  301422da869d749e5b4cab507366df366e74224c.

**Comment by yrliou** [comment]
Android build is broken, please fix it, see https://ci.brave.com/blue/organizations/jenkins/pr-brave-browser-f-wallet-jupiter-desktop-android/detail/pr-brave-browser-f-wallet-jupiter-desktop-android/18/pipeline.

**Review by yrliou — APPROVED** [review]
c++ backend LGTM, please create an issue for supporting partial_sign (existed signatures provided in the encoded transaction) for transactions coming from `MakeTxDataFromBase64EncodedTransaction`, and make sure Android build is green (might need to rebase to latest master to avoid new calls not being caught.).

**Comment by yrliou** [comment]
One more thing, please make sure to have security team's approval before merging.

**Comment by onyb** [comment]
@diracdeltas @yrliou Please review 84263d91e96250791b177a939efe6a78b634e69c which adds a new helper `convert_all_numbers_to_string` to the rust json lib. It recursively converts all JS numbers in the JSON to a string, so there's no requirement to provide an RFC6901 path unlike other helpers in the same module.

Consequently, we now have the following benefits:
1. We no longer perform C++ JSON parsing in the conversion callback, which is handled entirely in Rust.
2. Fields previously unreachable by the Rust `json` lib for conversion (`lpFee->amount` and `platformFee->amount`) are parsed the same way as any other field.

**Comment by yrliou** [comment]
Add @rillian and @antonok-edm for either of them to review the rust impl in https://github.com/brave/brave-core/commit/84263d91e96250791b177a939efe6a78b634e69c.

File: components/json/rs/src/lib.rs
Comment (antonok-edm) [inline components/json/rs/src/lib.rs:225]: I recommend triple-slash comments (`/// ...`) on function headers, as those will automatically be recognized as doc comments

**Review by antonok-edm — APPROVED** [review]

**Review by yrliou — APPROVED** [review]

**Review by diracdeltas — APPROVED** [review]

File: components/json/rs/src/lib.rs
Comment (bridiver) [inline components/json/rs/src/lib.rs:250]: Why are you doing this recursively? This is very inefficient. 
Also why are we applying this to everything instead of just applying it to the specific paths that we know need to be converted? This seems potentially dangerous to me and I think we should be much more strict about how we do conversions here.

**Pull request merged**
